File: PaintHouseRecursive.py



class Solution:
    def paint(self, costs) -> int:
        if(costs==None): return 0

        case1= self.helper(costs,0,0,0)
        case2= self.helper(costs,0,1,0)
        case3= self.helper(costs,0,2,0)

        return min(case1,min(case2,case3))

    def helper(self,costs,row,color,costsofar):
        if row == len(costs):
            return costsofar
        if color==0:
            return min( self.helper(costs,row+1,1,costsofar+costs[row][0]),
            self.helper(costs,row+1,2,costsofar+costs[row][0]))
        if color ==1:
            return min(self.helper(costs,row+1,0,costsofar + costs[row][1]),
            self.helper(costs,row+1,2,costsofar + costs[row][1]))
        if color ==2:
            return min(self.helper(costs,row+1,0,costsofar+costs[row][2]),
            self.helper(costs,row+1,1,costsofar + costs[row][2]))
        
costs = [[17,2,17],[16,16,5],[14,3,19]]
paint = Solution()
cost =paint.paint(costs)
print(cost)

File: paintHouseRecursive.java
class Solution {
    public int minCost(int[][] costs) {
        if (costs == null || costs.length == 0) {
            return 0;
        }
        
        int redCost = helper(costs, 0, 0, 0);
        int greenCost = helper(costs, 1, 0, 0);
        int blueCost = helper(costs, 2, 0, 0);
        return (Math.min(redCost, Math.min(greenCost, blueCost)));
    }
    
    private int helper(int costs[][], int color, int row, int totalCost) {
        //base
        if (row >= costs.length) {
            return totalCost;
        }
        
        // logic
        if (color == 0) {
            return Math.min(
                helper(costs, 1, row + 1, totalCost + costs[row][0]),
                helper(costs, 2, row + 1, totalCost + costs[row][0])
            );
        } else if (color == 1) {
            return Math.min(
                helper(costs, 0, row + 1, totalCost + costs[row][1]),
                helper(costs, 2, row + 1, totalCost + costs[row][1])
            );
        } else {
            return Math.min(
                helper(costs, 0, row + 1, totalCost + costs[row][2]),
                helper(costs, 1, row + 1, totalCost + costs[row][2])
            );
        }
    }
}